src/api_server/src/lib.rs

```diff
@@ -172,13 +185,45 @@ impl ApiServer {
         }
     }
 
-    fn serve_vmm_action_request(&self, vmm_action: Box<VmmAction>) -> Response {
+    fn serve_vmm_action_request(
+        &self,
+        vmm_action: Box<VmmAction>,
+        action_start_us: u64,
+    ) -> Response {
+        let metric_with_action = match *vmm_action {
+            #[cfg(target_arch = "x86_64")]
+            VmmAction::CreateSnapshot(ref params) => match params.snapshot_type {
+                SnapshotType::Full => Some((
+                    &METRICS.latencies_us.full_create_snapshot,
+                    "creating the full snapshot",
+                )),
+                SnapshotType::Diff => Some((
+                    &METRICS.latencies_us.diff_create_snapshot,
+                    "creating the diff snapshot",
+                )),
+            },
+            #[cfg(target_arch = "x86_64")]
+            VmmAction::LoadSnapshot(_) => {
+                Some((&METRICS.latencies_us.load_snapshot, "loading the snapshot"))
+            }
+            VmmAction::Pause => Some((&METRICS.latencies_us.pause_vm, "pausing the microVM")),
+            VmmAction::Resume => Some((&METRICS.latencies_us.resume_vm, "resuming the microVM")),
+            _ => None,
+        };
+
         self.api_request_sender
             .send(vmm_action)
             .expect("Failed to send VMM message");
         self.to_vmm_fd.write(1).expect("Cannot update send VMM fd");
         let vmm_outcome = *(self.vmm_response_receiver.recv().expect("VMM disconnected"));
-        ParsedRequest::convert_to_response(vmm_outcome)
+        let response = ParsedRequest::convert_to_response(&vmm_outcome);
+
+        if vmm_outcome.is_ok() {
+            if let Some((metric, action)) = metric_with_action {
+                update_metric_with_elapsed_time(metric, action_start_us, action, "API");
```

Comment:
I'm not sure about the added value of the `update_metric_with_elapsed_time()` function.

```suggestion
                let elapsed = utils::elapsed_us(action_start_us);
                metric.store(elapsed);
                debug!("API thread took {} us for {:?}", elapsed, vmm_action);
```

You get a couple duplicated lines but it's easier to read and maintain (less bytes used too).

Avoids the `action` static string description and only formats `vmm_action` when log-level is debug.

Reply:
I agree it's much clearer without the separate function that does all that stuff, I actually wanted to separate the logging part and add a function in a `SharedMetric` implementation: `update_with_elapsed_time(&self, start_time)` in order to avoid the ugly metric parameter.
This duplicate code happens six times, so I thought it deserves a separate function though (and not pollute the ones from the vmm crate with all these logs/metrics).
I think I'll leave `update_metric_with_elapsed_time(metric, action_start_us)` outside of any `SharedMetric` implementation and return the elapsed time (and the logging part will be done in `vmm`); sounds like a good compromise solution to me :-?.

Reply:
I thought about `update_metric_with_elapsed_time(metric, action_start_us)` to return elapsed time as well, but in the end I think the function is still weird.

```rust
let elapsed = utils::elapsed_us(action_start_us);
metric.store(elapsed);
```
does the same thing in a nice way keeping separation of concerns, vs.
```rust
let elapsed = update_metric_with_elapsed_time(metric, action_start_us);
```

The second option is also _fine_ if you insist on it, not gonna block the PR for it.

Reply:
`let elapsed = SOME_METRIC.update_with_elapsed_time(action_start_us);` is definitely weird and wanted to avoid. Option 2 seems ok to me.